// ngx-auth-firebaseui/esm2015/lib/components/ngx-auth-firebaseui-avatar/ngx-auth-firebaseui-avatar.component.js
import { __awaiter, __decorate } from "tslib";
import { Component, EventEmitter, Input, Output } from '@angular/core';
import { AngularFireAuth } from '@angular/fire/auth';
import { MatDialog } from '@angular/material/dialog';
import { UserComponent } from '..';
let NgxAuthFirebaseuiAvatarComponent = class NgxAuthFirebaseuiAvatarComponent {
    constructor(afa, dialog) {
        this.afa = afa;
        this.dialog = dialog;
        this.canLogout = true;
        this.canDeleteAccount = true;
        this.canEditAccount = true;
        // tslint:disable-next-line:no-output-on-prefix
        this.onSignOut = new EventEmitter();
    }
    ngOnInit() {
        this.user$ = this.afa.user;
        this.user$.subscribe((user) => {
            this.user = user;
            this.displayNameInitials = user ? this.getDisplayNameInitials(user.displayName) : null;
        });
    }
    getDisplayNameInitials(displayName) {
        if (!displayName) {
            return null;
        }
        const initialsRegExp = displayName.match(/\b\w/g) || [];
        const initials = ((initialsRegExp.shift() || '') + (initialsRegExp.pop() || '')).toUpperCase();
        return initials;
    }
    openProfile() {
        const dialogRef = this.dialog.open(UserComponent);
        const instance = dialogRef.componentInstance;
        instance.canDeleteAccount = this.canDeleteAccount;
        instance.canEditAccount = this.canEditAccount;
    }
    signOut() {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                yield this.afa.signOut();
                // Sign-out successful.
                this.onSignOut.emit();
            }
            catch (e) {
                // An error happened.
                console.error('An error happened while signing out!', e);
            }
        });
    }
};
NgxAuthFirebaseuiAvatarComponent.ctorParameters = () => [
    { type: AngularFireAuth },
    { type: MatDialog }
];
__decorate([
    Input()
], NgxAuthFirebaseuiAvatarComponent.prototype, "canLogout", void 0);
__decorate([
    Input()
], NgxAuthFirebaseuiAvatarComponent.prototype, "links", void 0);
__decorate([
    Input()
], NgxAuthFirebaseuiAvatarComponent.prototype, "canDeleteAccount", void 0);
__decorate([
    Input()
], NgxAuthFirebaseuiAvatarComponent.prototype, "canEditAccount", void 0);
__decorate([
    Output()
], NgxAuthFirebaseuiAvatarComponent.prototype, "onSignOut", void 0);
NgxAuthFirebaseuiAvatarComponent = __decorate([
    Component({
        selector: 'ngx-auth-firebaseui-avatar',
        template: "<button *ngIf=\"user\"\n        [matMenuTriggerFor]=\"posXMenu\"\n        [matTooltip]=\"user?.displayName\"\n        [style.background-image]=\"'url(' + user?.photoURL + ')'\"\n        aria-label=\"Open x-positioned menu\"\n        mat-mini-fab\n        style=\"background-size: cover\">\n  <span *ngIf=\"!user?.photoURL\">{{displayNameInitials || ''}}</span>\n</button>\n\n<mat-menu #posXMenu=\"matMenu\" class=\"before\" xPosition=\"before\">\n  <div fxLayout=\"row\" fxLayout.xs=\"column\" style=\"padding-left: 10px; padding-right: 10px\">\n    <button [style.background-image]=\"'url(' + user?.photoURL + ')'\"\n            mat-fab\n            style=\"background-size: cover\">\n      <span *ngIf=\"!user?.photoURL\">{{displayNameInitials || ''}}</span>\n    </button>\n    <div fxLayout=\"column\" style=\"padding-left: 10px; padding-right: 10px\">\n      <strong mat-card-title>{{user?.displayName}}</strong>\n      <em mat-card-subtitle style=\"font-style: italic\">{{user?.email}}</em>\n    </div>\n  </div>\n\n  <div fxFlex=\"100\" fxLayout=\"column\">\n    <div *ngFor=\"let menuItem of links\" class=\"links-menu\">\n      <button (click)=\"menuItem?.callback()\" mat-menu-item>\n        <mat-icon>{{menuItem?.icon}}</mat-icon>\n        {{menuItem?.text}}</button>\n    </div>\n    <button (click)=\"openProfile()\" color=\"primary\" fxLayoutAlign=\"center\" mat-raised-button>Profile\n    </button>\n    <button (click)=\"signOut()\" *ngIf=\"canLogout\" color=\"warn\" fxLayoutAlign=\"center\" mat-raised-button>Sign Out\n    </button>\n  </div>\n</mat-menu>\n",
        styles: [".mat-raised-button{margin:.2rem 1rem}.links-menu{text-align:center}"]
    })
], NgxAuthFirebaseuiAvatarComponent);
export { NgxAuthFirebaseuiAvatarComponent };
